boards/common/blxxxpill/Makefile.include

```diff
@@ -7,6 +7,23 @@ ifeq (dfu-util,$(PROGRAMMER))
 
   # Leave some space at the beginning of the flash for the bootloader
   ROM_OFFSET ?= 0x2000
+
+  # If using STM32duino bootloader, this change is necessary.
+  # Bootloader available at
+  # github.com/rogerclarkmelbourne/STM32duino-bootloader/tree/master/binaries
+  ifeq (stm32duino,$(BOOTLOADER))
+    FFLAGS_OPTS ?= -R
+    FFLAGS ?= --device $(DFU_USB_ID) \
+              --alt $(DFU_ALT) \
+              --download $(FLASHFILE) \
+              $(FFLAGS_OPTS)
+
+    # Flashing may be easier when uncommenting the "RESET_STM" line and
+    # using the binary "upload-reset" from rogerclarkmelbourne/Arduino_STM32.
+    # github.com/rogerclarkmelbourne/Arduino_STM32/tree/master/tools/linux/upload-reset
+    # RESET_STM := $(shell $(RIOTBOARD)/common/blxxxpill/bin/upload-reset /dev/ttyACM0 750)
+  endif
+
```

Comment:
This should be documented in `boards/common/blxxxpill/doc.txt`. (@benpicco: The other dfu bootloader doc is btw also missing. Care to add this?)

Also: The README of the stm32duino bootloader says this:

> Also Note: Use GCC 4.8 (not 4.9 or newer, as those versions have more aggressive optimisation which cause hardware registers to be read incorrectly and consequently the bootloader does not work)

This is a highly significant indicator of code quality problems of the bootloader. (If your code breaks by updating GCC, in 999 out of 1000 cases the problem is your code, not GCC. I bet they forgot to add the `volatile` qualifier for accessing memory mapped I/O or similar.) Maybe it is just better to use a different bootloader instead?

Reply:
I think you did put a very pertinent question.

I got to know Riot a few years ago, but haven`t worked on it so far. Now I got some cheap boards (bluepills) to experience this system as an alternative or my projects. In my short experience working on STM32, I went through a lot of options to flash it, and one of the most cited/documented in the community was this **stm32duino bootloarder**.

During my attempts to flash using dfu provided in Arduino and Riot, I found this difference that made it actually possible to use their bootloader into Riot. With Riot`s default configuration for dfu, it did not work at all.

So, in general, I think you made a very good point, @benpicco  indeed could give us some insight, but I think that leaving this `BOOTLOADER=stm32duino` option might be very useful for newcomers like me, especially if they have lesser knowledge about these things.

Reply:
Sorry I've bee busy the last couple of days - I'm not sure why you assume I have much knowledge about DFU bootloaders though :laughing: 
You added the config for `dfu-util`(#12697), I just happened to use it. So I'm also not sure what doc you mean.

If using `upload-reset` makes flashing easier, why not always use it and download that tool on-demand? I wouldn't download the binary though, just the source and compile it like it's done with other [tools](https://github.com/RIOT-OS/RIOT/tree/master/dist/tools).

`upload-reset` would then be added to `FLASHDEPS`.

Reply:
Is the `RESET_STM` variable anywhere referenced? It seems to neither be referenced in code introduced by this PR nor by existing code in RIOT. Thus, I don't see how this should have an impact.

Reply:
```
    # Flashing may be easier when uncommenting the "RESET_STM" line and
    # using the binary "upload-reset" from rogerclarkmelbourne/Arduino_STM32.
    # github.com/rogerclarkmelbourne/Arduino_STM32/tree/master/tools/linux/upload-reset
    # RESET_STM := $(shell $(RIOTBOARD)/common/blxxxpill/bin/upload-reset /dev/ttyACM0 750)
```
`RESET_STM` is commented as one possible solution/improvement for someone trying stm32duino bootloader.

Reply:
But how would it be used? `RESET_STM` is not called anywhere.

Reply:
Oh, I misunderstood the question.
`upload-reset` is called as soon as the `RESET_STM` variable is assigned because of `:=`

`:= $(shell`

Reply:
Uh yea! That's easy to overlook.

I think you want to use 

    PREFLASHER = /path/to/upload-reset
    PREFFLAGS = $(PORT) 750

here.

Reply:
Oh ok. I did not see this PREFLASHER option before. Nice!
Ok, it could be changed to that, IF your first USB reset does not work instead since we already have that.
But in any case, this is only useful if USB is **on**, which is not yet. I don`t know if it`s worth changing this right now. I gotta check the USB first.

Reply:
Yea so better remove it for now, maybe add a comment that `upload-reset` can be used in the future.

Reply:
I still think that a comment would still be helpful for some people and still not affect code anyhow. But, in this case, it's your call.

Reply:
A comment that `upload-reset /dev/ttyACM0 750` should be used as `PREFLASHER` once we have USB support for stm32f1 is good, but let's not get people to think this implicit reset at Makefile evaluation would be a good idea :wink:

Reply:
Doing it now. But, important to remember that the bootloader has USB support, and that is why the reset command can improve this procedure.
So I`m going to test this `PREFLASHER` option now.